### PR TITLE
*: enable process collector and add thread collector

### PR DESCRIPTION
**Pull request by overvenus**
This PR adds a custom prometheus collector, it collects each thread's CPU time. Also this PR updates promethues crate and enables the `process` feature.

PTAL @zhangjinpeng1987 @siddontang @BusyJay 

Close #1481

### CODE REVIEW

File: src/util/mod.rs [hunk 1]
```diff
@@ -376,21 +377,34 @@ pub fn run_prometheus(interval: Duration,
...
-            let metric_familys = prometheus::gather();
-
-            let res = prometheus::push_metrics(&job,
-                                               prometheus::hostname_grouping_key(),
-                                               &address,
-                                               metric_familys);
-            if let Err(e) = res {
-                error!("fail to push metrics: {}", e);
+    let handler = thread::Builder::new()
+        .name("PromePusher".to_owned())
```
Comment (siddontang) [inline src/util/mod.rs]: use lowercase for the thread name.

File: src/util/thread_metrics.rs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            Some(tid) => try!(tid.parse().map_err(|e| to_err(format!("{}", e)))),
+            None => return Err(to_err(format!("fail to read {} task", pid))),
+        };
+
+        let mut tname = String::new();
+        try!(fs::File::open(format!("/proc/{}/task/{}/comm", pid, tid))
+            .and_then(|mut f| f.read_to_string(&mut tname)));
+
+        // Pop tail '\n'.
+        tname.pop();
```
Comment (siddontang) [inline src/util/thread_metrics.rs]: can we use trim space? I worry the tail is not `\n` sometimes.

File: src/util/thread_metrics.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+impl Collector for ThreadsColletcor {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
+        let mut core = self.core.lock().unwrap();
+        core.refresh += 1;
+        if core.refresh >= REFRESH_CNT {
```
Comment (siddontang) [inline src/util/thread_metrics.rs]: why do we need refreshing here? Getting threads every time in collect can cause performance reduction?

File: src/util/thread_metrics.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+            'a'...'z' | 'A'...'Z' | '0'...'9' | '_' | ':' => {
+                san.push(c);
+            }
+            '-' => san.push('_'),
+            _ => (),
+        }
+    }
+
+    if san.is_empty() {
+        format!("unkonw_{}", TH_CNT.fetch_add(1, Ordering::Relaxed))
```
Comment (siddontang) [inline src/util/thread_metrics.rs]: can we meet a thread with no name?

Comment (overvenus) [inline src/util/thread_metrics.rs]: `"%^"` -> `""`, filtered by line [133](https://github.com/pingcap/tikv/pull/1494/files/04f5bf5f1d1a3959c1482d725b7442f9ac7a7e69#diff-bcd20d7f9f1de0c9c51cc4573cc54c06R133)

Comment (siddontang) [inline src/util/thread_metrics.rs]: unknown

Comment (siddontang) [inline src/util/thread_metrics.rs]: the unknown thread may use different counter when call sanitize_thread_name multi times.

File: src/util/thread_metrics.rs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
+        let cpu_totals = self.cpu_totals.lock().unwrap();
+        let pairs = get_threads(self.pid).unwrap();
+        for (tid, tname) in pairs {
+            if let Ok((utime, stime)) = find_thread_cpu_time(self.pid, tid) {
+                let total = (utime + stime) / *CLK_TCK;
+                let cpu_total = cpu_totals.get_metric_with_label_values(&[&tname]).unwrap();
```
Comment (siddontang) [inline src/util/thread_metrics.rs]: does tname always exist?

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

File: src/util/thread_metrics.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: Happy new year!

File: src/util/thread_metrics.rs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        san
+    }
+}
+
+fn to_err(s: String) -> Error {
+    Error::new(ErrorKind::Other, s)
+}
+
+// See more man proc.
+const CPU_INDEX: [usize; 2] = [14 - 1, 15 - 1];
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: What if file name contains space?

File: src/util/thread_metrics.rs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<proto::MetricFamily> {
+        let cpu_totals = self.cpu_totals.lock().unwrap();
+        let pairs = get_threads(self.pid).unwrap();
+        for (tid, tname) in pairs {
+            if let Ok((utime, stime)) = find_thread_cpu_time(self.pid, tid) {
+                let total = (utime + stime) / *CLK_TCK;
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: Why divide CLK_TCK?

File: src/util/thread_metrics.rs [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+    let dirs = try!(fs::read_dir(format!("/proc/{}/task", pid)));
+    for task in dirs {
+        let tid = match try!(task).file_name().to_str() {
+            Some(tid) => try!(tid.parse().map_err(|e| to_err(format!("{}", e)))),
+            None => return Err(to_err(format!("fail to read {} task", pid))),
+        };
+
+        let mut tname = String::new();
+        try!(fs::File::open(format!("/proc/{}/task/{}/comm", pid, tid))
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: You can read thread name from "/proc/pid/task/tid/stat" directly.

File: src/util/thread_metrics.rs [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    }
+}
+
+fn get_threads(pid: pid_t) -> Result<Vec<(pid_t, String)>> {
+    let mut pairs = Vec::new();
+
+    let dirs = try!(fs::read_dir(format!("/proc/{}/task", pid)));
+    for task in dirs {
+        let tid = match try!(task).file_name().to_str() {
+            Some(tid) => try!(tid.parse().map_err(|e| to_err(format!("{}", e)))),
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: `continue` instead.

File: src/util/thread_metrics.rs [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            None => return Err(to_err(format!("fail to read {} task", pid))),
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: /{} task/task of {}/

File: src/util/thread_metrics.rs (hunk 6 above)
Comment (overvenus) [inline src/util/thread_metrics.rs]: And to you :)

File: src/util/thread_metrics.rs (hunk 8 above)
Comment (overvenus) [inline src/util/thread_metrics.rs]: So we get time, measured in seconds.

> (14) utime  %lu
>                        Amount of time that this process has been scheduled
>                        in user mode, measured in clock ticks (divide by
>                        sysconf(_SC_CLK_TCK)).

Comment (BusyJay) [inline src/util/thread_metrics.rs]: Does it mean the value was divided by sysconf(_SC_CLK_TCK) already?

Comment (overvenus) [inline src/util/thread_metrics.rs]: No, `tick` is the actually unit of those nembers. I think it means:

    tick = second / sysconf(_SC_CLK_TCK)

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/util/thread_metrics.rs [hunk 12]
```diff
@@ -78,62 +78,90 @@ impl Collector for ThreadsColletcor {
...
-fn sanitize_thread_name(tid: pid_t, tname: String) -> String {
-    let mut san = String::with_capacity(tname.len());
-    for c in tname.chars() {
+fn get_thread_name(tid: pid_t, stat: &str) -> (String, usize) {
+    let mut space_cnt = 0;
+    let mut open_brkt = 0;
+    let mut name = String::with_capacity(THD_NAME_LEN);
+    for c in stat.chars() {
         match c {
+            // TODO: properly handle `)a(`.
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: what if the name is ")a S"? A better way to handle this is treat substring between first '(' and last ')' as the thread's name.

File: src/util/thread_metrics.rs [hunk 13]
```diff
@@ -78,62 +78,90 @@ impl Collector for ThreadsColletcor {
...
 
-    let stats: Vec<_> = buffer.split_whitespace().collect();
-    let utime = stats[CPU_INDEX[0]];
-    let stime = stats[CPU_INDEX[1]];
-    Ok((utime.parse().unwrap(), stime.parse().unwrap()))
+// Extracted from `get_thread_stat`, for test purpose.
+fn get_thread_stat_internal(tid: pid_t, stat: &str) -> (String, f64, f64) {
+    let (name, space_cnt) = get_thread_name(tid, stat);
+    let stats: Vec<_> = stat.split_whitespace().collect();
+    let utime = stats[CPU_INDEX[0] + space_cnt];
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: This is a little tricky. You can only handle the substring of stat which doesn't contains pid and pthread name.

File: src/util/thread_metrics.rs [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+        cpu_totals.collect()
+    }
+}
+
+fn get_thread_ids(pid: pid_t) -> Result<Vec<pid_t>> {
+    let mut tids = Vec::new();
+    let dirs = try!(fs::read_dir(format!("/proc/{}/task", pid)));
+    for task in dirs {
+        let tid = match try!(task).file_name().to_str() {
+            Some(tid) => try!(tid.parse().map_err(|e| to_err(format!("{}", e)))),
```
Comment (BusyJay) [inline src/util/thread_metrics.rs]: Should continue too.

**Comment by BusyJay** [comment]
Ping

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**